gqcp/include/Mathematical/Grid/CubicGrid.hpp

```diff
@@ -0,0 +1,121 @@
+// This file is part of GQCG-gqcp.
+// 
+// Copyright (C) 2017-2019  the GQCG developers
+// 
+// GQCG-gqcp is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+// 
+// GQCG-gqcp is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Lesser General Public License for more details.
+// 
+// You should have received a copy of the GNU Lesser General Public License
+// along with GQCG-gqcp.  If not, see <http://www.gnu.org/licenses/>.
+// 
+#pragma once
+
+
+#include "Mathematical/Grid/Field.hpp"
+#include "Mathematical/Representation/Matrix.hpp"
+#include "Mathematical/ScalarFunction.hpp"
+
+#include <array>
+#include <functional>
+#include <type_traits>
+
+
+namespace GQCP {
+
+
+/**
+ *  A grid type whose points are on a regular cubic lattice.
+ */
+class CubicGrid {
+private:
+    Vector<double, 3> m_origin;  // the origin of the grid
+    std::array<size_t, 3> m_steps;  // the number of steps in the x, y, z-directions
+    std::array<double, 3> step_sizes;  // the step sizes in the x, y, z-directions
+
+
+public:
+    // CONSTRUCTORS
+
+    /**
+     *  @param origin               the origin of the grid
+     *  @param steps                the number of steps in the x, y, z-directions
+     *  @param step_sizes           the step sizes in the x, y, z-directions
+     */
+    CubicGrid(const Vector<double, 3>& origin, const std::array<size_t, 3>& steps, const std::array<double, 3>& step_sizes);
+
+
+    // PUBLIC METHODS
+
+    /**
+     *  Evaluate a scalar function on every point of this grid.
+     * 
+     *  @param scalar_function          the scalar functions whose values should be evaluated
+     * 
+     *  @return a field with the calculated evaluations
+     */
+    template <typename Valued>
+    Field<Valued, CubicGrid> evaluate(const ScalarFunction<Valued, double, 3>& scalar_function) const {
+
+        std::vector<Valued> values;  // the evaluated values of the scalar function
+        values.reserve(this->numberOfPoints());
+
+        this->loop( [&values, &scalar_function] (const Vector<double, 3>& r) {
+            values.push_back(scalar_function(r));
+        });
+
+        return Field<Valued, CubicGrid>{values, *this};
+    }
+
+    /**
```

Comment:
```suggestion
    }


    /**
```
We've always used two spaces between methods.